azure-iot-device/azure/iot/device/common/http_transport.py

```diff
@@ -0,0 +1,193 @@
+    during instantation of the transport, in which case the HTTPSConnection would be shared among all
+    requests! Cool? Maybe. I don't know precisely, but I'm pretty sure that if the conneciton is making a
+    request, and then another thread comes along and using the same connection makes a request... well I just
+    don't really know what's going to happen. So the second (and currently implemented) option for this is to
+    create a new HTTPSConnection object every time we make a request, so that each request has it's own "connection".
+    What does this mean? Not sure, but it seem safer.
+
+    2) Initially I atomized each http_client call. We would create the client in one call, we would connect in another,
+    disconnect in yet another. Seemed fine... ish. But then Carter was like hey we should make all the http_client specific
+    stuff down here and only expose a relevant surface. What's a relevant surface for HTTP? Probably just request. So that's
+    what I'm exposing. Everything, the insantiation of the HTTPSConnection, the sending of the request, the getting of the response,
+    all happens in this one call, and that's better. Probably.
+    """
+
+    @pipeline_thread.invoke_on_http_thread_nowait
+    # TODO: This star syntax is incompatible with Python 2, change it so that the callback is in front of the optional params.
+    def request(self, method, hostname, path, callback, body=None, headers={}, query_params=None):
+        # Sends a complete request to the server
+        logger.info("sending https request")
+        try:
+            logger.debug("creating to https connection")
+            connection = http_client.HTTPSConnection(
+                self._hostname, context=self._create_ssl_context()
+            )
+            logger.debug("connecting to host tcp socket")
+            connection.connect()
+            logger.debug("connection succeeded")
+            # formatted_headers = self._format_headers(headers)
+            url = "https://{hostname}/{path}?{query_params}".format(
+                hostname=hostname, path=path, query_params=query_params
+            )
+            connection.request(method, url, body=body.encode("utf-8"), headers=headers)
+            response = connection.getresponse()
+            status_code = response.status
+            response_string = response.read()
+
+            logger.debug("response received.")
+            logger.debug("closing connection to https host")
+            connection.close()
+            logger.debug("connection closed")
+            logger.info("https request sent")
+            callback(status_code, response_string)
+        except Exception as e:
+            # TODO: This exception needs to be returned in the callback, I cannot raise here because it's in a different thread than the pipeline
+            # and would therefore do nothing. Instead pass an exception to the callback as an error if this is excepted.
```

Comment:
I don't think this is true?

Reply:
If we're looking for parity, the MQTT transport is not a good standard to base ourselves on.  Paho has 3 different ways to return errors: an API can raise, an API can return a filing RC code, and _some_ callbacks can return failures.  
No matter how we do threading in the HTTP transport, we will _always_ need a way to return some errors via callback.  Forcing some errors to be raised in the invocation thread feels like parity for the sake of parity without any real benefit.

---
In reply to: [351480077](https://github.com/Azure/azure-iot-sdk-python/pull/403#discussion_r351480077) [](ancestors = 351480077)

Reply:
The threading model used in MQTT shouldn't exactly be used to compare this since what I'm implementing is the equivalent of Paho and the MQTT Transport, not just the MQTT Transport. But without debating that, raising an error here would be confusing and less than ideal. Passing it up as a callback parameter is necessary if this is running on it's own thread asynchronously.

---
In reply to: [352765989](https://github.com/Azure/azure-iot-sdk-python/pull/403#discussion_r352765989) [](ancestors = 352765989,351480077)